subsys/net/lib/nrf_cloud/coap/src/nrf_cloud_coap_transport.c

```diff
@@ -0,0 +1,457 @@
+	};
+	char ipv4_addr[NET_IPV4_ADDR_LEN];
+
+	LOG_DBG("Looking up server %s", CONFIG_NRF_CLOUD_COAP_SERVER_HOSTNAME);
+	err = getaddrinfo(CONFIG_NRF_CLOUD_COAP_SERVER_HOSTNAME, NULL, &hints, &result);
+	if (err != 0) {
+		LOG_ERR("ERROR: getaddrinfo for %s failed: %d",
+			CONFIG_NRF_CLOUD_COAP_SERVER_HOSTNAME, err);
+		return -EIO;
+	}
+
+	if (result == NULL) {
+		LOG_ERR("ERROR: Address not found");
+		return -ENOENT;
+	}
+
+	/* IPv4 Address. */
+	struct sockaddr_in *server4 = ((struct sockaddr_in *)&server);
+
+	server4->sin_addr.s_addr =
+		((struct sockaddr_in *)result->ai_addr)->sin_addr.s_addr;
+	server4->sin_family = AF_INET;
+	server4->sin_port = htons(CONFIG_NRF_CLOUD_COAP_SERVER_PORT);
+
+	connection_info.s4_addr[0] = server4->sin_addr.s4_addr[0];
+	connection_info.s4_addr[1] = server4->sin_addr.s4_addr[1];
+	connection_info.s4_addr[2] = server4->sin_addr.s4_addr[2];
+	connection_info.s4_addr[3] = server4->sin_addr.s4_addr[3];
+
+	inet_ntop(AF_INET, &server4->sin_addr.s_addr, ipv4_addr,
+		  sizeof(ipv4_addr));
+	LOG_INF("Server %s IP address: %s, port: %u",
+		CONFIG_NRF_CLOUD_COAP_SERVER_HOSTNAME, ipv4_addr,
+		CONFIG_NRF_CLOUD_COAP_SERVER_PORT);
+
+	/* Free the address. */
+	freeaddrinfo(result);
+
+	return 0;
+}
+
+/**@brief Initialize the CoAP client */
+int nrf_cloud_coap_init(void)
+{
+	static bool initialized;
+	int err;
+
+	authorized = false;
+
+	if (!initialized) {
+		/* Only initialize one time; not idempotent. */
+		LOG_INF("Initializing async coap client");
+		err = coap_client_init(&coap_client, NULL);
+		if (err) {
+			LOG_ERR("Failed to initialize coap client: %d", err);
+			return err;
+		}
+		(void)nrf_cloud_codec_init(NULL);
+		initialized = true;
+	}
+
+	return 0;
+}
+
+int nrf_cloud_coap_connect(void)
+{
+	int err;
+
+	err = server_resolve();
+	if (err) {
+		LOG_ERR("Failed to resolve server name: %d", err);
+		return err;
+	}
+
+	LOG_DBG("Creating socket type IPPROTO_DTLS_1_2");
+	sock = socket(AF_INET, SOCK_DGRAM, IPPROTO_DTLS_1_2);
+	LOG_DBG("sock = %d", sock);
+	if (sock < 0) {
+		LOG_ERR("Failed to create CoAP socket: %d.", -errno);
+		return -errno;
+	}
+
+	err = dtls_init(sock);
+	if (err < 0) {
+		LOG_ERR("Failed to initialize the DTLS client: %d", err);
+		return err;
+	}
+
+	if (dtls_cid_is_available()) {
+		err = dtls_session_load(sock);
+		if (!err) {
+			LOG_INF("  Loaded DTLS CID session");
+			authorized = true;
+		} else {
+			LOG_INF("  No DTLS CID session loaded: %d", err);
+		}
+	} else {
+		LOG_INF("  DTLS CID is not available");
+	}
+
+	/* Note: connect() of a SOCK_DGRAM socket does not actually connect, because
+	 * UDP is connectionless. The first actual traffic, including DTLS
+	 * handshake if needed, occurs when a sendto() or recvfrom() occurs on the
+	 * socket.
+	 */
+	err = connect(sock, (struct sockaddr *)&server,
+		      sizeof(struct sockaddr_in));
+	if (err < 0) {
+		LOG_ERR("Connect failed : %d", -errno);
+		return -errno;
+	}
+
+	return nrf_cloud_coap_authorize();
```

Comment:
This function is missing cleanup of the socket if `nrf_cloud_coap_authorize` fails. I had a few subsequent calls to connect (failing for some reason), where I eventually run out of sockets.

Reply:
good catch